component/azstorage/block_blob.go

```diff
@@ -653,6 +656,179 @@ func (bb *BlockBlob) WriteFromBuffer(name string, metadata map[string]string, da
+		// update the data with the new data
+		// if we're only overwriting existing data
+		if int64(len(oldData)) >= offset+length {
+			copy(oldData[offset:], data)
+			dataBuffer = &oldData
+			// else appending and/or overwriting
+		} else {
+			// if the file is not empty then we need to combine the data
+			if len(oldData) > 0 {
+				// new data buffer with the size of old and new data
+				newDataBuffer := make([]byte, offset+length)
+				// copy the old data into it
+				// TODO: better way to do this?
+				if offset != 0 {
+					copy(newDataBuffer, oldData)
+					oldData = nil
+				}
+				// overwrite with the new data we want to add
+				copy(newDataBuffer[offset:], data)
+				dataBuffer = &newDataBuffer
+			} else {
+				dataBuffer = &data
+			}
+		}
+		// WriteFromBuffer should be able to handle the case where now the block is too big and gets split into multiple blocks
+		err := bb.WriteFromBuffer(name, nil, *dataBuffer)
+		if err != nil {
+			log.Err("BlockBlob::Write : Failed to upload to blob %s ", name, err.Error())
+			return err
+		}
+		// case 2: given offset is within the size of the blob - and the blob consists of multiple blocks
+		// case 3: new blocks need to be added
+	} else {
+		index, oldDataSize, exceedsFileBlocks, appendOnly := fileOffsets.FindBlocksToModify(offset, length)
+		// keeps track of how much new data will be appended to the end of the file (applicable only to case 3)
+		newBufferSize := int64(0)
+		// case 3?
+		if exceedsFileBlocks {
+			// get length of blockID in order to generate a consistent size block ID so storage does not throw
+			existingBlockId, _ := base64.StdEncoding.DecodeString(fileOffsets.BlockList[0].Id)
+			blockIdLength := len(existingBlockId)
+			newBufferSize = bb.createNewBlocks(fileOffsets, offset, length, int64(blockIdLength))
+		}
+		if !appendOnly {
+			// buffer that holds that pre-existing data in those blocks we're interested in
+			oldDataBuffer := make([]byte, oldDataSize+newBufferSize)
+			if !appendOnly {
+				// fetch the blocks that will be impacted by the new changes so we can overwrite them
+				bb.ReadInBuffer(name, fileOffsets.BlockList[index].StartIndex, oldDataSize, oldDataBuffer)
+			}
+			// this gives us where the offset with respect to the buffer that holds our old data - so we can start writing the new data
+			blockOffset := offset - fileOffsets.BlockList[index].StartIndex
+			copy(oldDataBuffer[blockOffset:], data)
+			err := bb.stageAndCommitModifiedBlocks(name, oldDataBuffer, index, fileOffsets)
+			return err
+		}
+	}
+	return nil
+}
+
+// TODO: make a similar method facing stream that would enable us to write to cached blocks then stage and commit
+func (bb *BlockBlob) stageAndCommitModifiedBlocks(name string, data []byte, index int, offsetList *common.BlockOffsetList) error {
+	blobURL := bb.Container.NewBlockBlobURL(filepath.Join(bb.Config.prefixPath, name))
+	blockOffset := int64(0)
+	for _, blk := range offsetList.BlockList[index:] {
```

Comment:
795 and 814 loop can be combined. Just iterate the entire list once. For each item append id to blockidList and if the block is modified then stage it in the same loop.

Reply:
This will also help us to reuse the exact same method when stream component kicks in as it will mostly send the blockIdList where multiple blocks are marked as modified and they may not be adjacent to each-other (random write case)

Reply:
done